### ISSUE TITLE
[CSA] False negative: a zero constraint returned by a trivial helper is lost

### ISSUE THREAD

**Issue by chisa22** [llvm/llvm-project#214225]
Hi, I found a false negative in `core.DivideZero`.

## Program

```cpp
int is_zero(unsigned value) {
  return value == 0;
}

void test(unsigned value) {
  if (!is_zero(value))
    return;
  (void)(7 / value);
}
```

The division is reached only when `is_zero(value)` returned true, so `value` is zero. CSA emits no warning. Replacing the condition with the equivalent direct predicate `if (value != 0) return;` produces the expected `core.DivideZero` result.

## CSA reproduction

```sh
clang++ --analyze \
  -Xclang -analyzer-checker=core.DivideZero \
  helper_return_constraint_divide_zero.cpp
```

## Comparison with related programs

The direct predicate form is reported:

```cpp
void test(unsigned value) {
  if (value != 0)
    return;
  (void)(7 / value); // warning: division by zero
}
```

The missed program wraps the same zero predicate in a trivial helper:

```cpp
int is_zero(unsigned value) {
  return value == 0;
}

void test(unsigned value) {
  if (!is_zero(value))
    return;
  (void)(7 / value); // no warning
}
```

## Runtime confirmation

Executing the program with Undefined BehaviorSanitizer reports division by zero.

## Expected result

Inlining a side-effect-free boolean helper should propagate its return constraint to the caller's subsequent path.

## Verification

```text
clang version 24.0.0git (https://github.com/llvm/llvm-project.git 1cb7e838cd47ecad4050948c0c907ecb1f466ac3)
```

**Comment by NagyDonat** [comment]
I verified that the report is created but then it is suppressed by the logic implemented in [`SuppressInlineDefensiveChecksVisitor::VisitNode
`](https://clang.llvm.org/doxygen/BugReporterVisitors_8cpp_source.html#l01850).

This suppression heuristic is indispensable because without it the analyzer would generate lots and lots of false positives when it encounters defensive checks like `if (ptr == nullptr) return;` at the beginning of functions. (By default the logic of the analyzer is that the presence of a `ptr == nullptr` check implies that it is worth to investigate the `ptr == nullptr` case separately -- but defensive programmers often insert such checks even in cases where `ptr` cannot actually be a nullpointer, so the analyzer must discard reports based on such unjustified assumptions.)

The reported false negative is very similar to the common false positive pattern that needs to stay suppressed, so I don't see a clear way to resolve this bug. Perhaps we need to just accept that this buggy code cannot be automatically distinguished from innocent code.

@steakhal @Xazax-hun @ anybody else 
Do you happen to have an idea to tweak this suppression heuristic in a way that improves its accuracy?

**Comment by steakhal** [comment]
It deserves careful thinking that I don't really have the time for right now. I don't think I have anything to add.
